rules/DowngradePhp80/Rector/ClassMethod/DowngradeAbstractPrivateMethodInTraitRector.php

```diff
@@ -0,0 +1,72 @@
+<?php
+
+declare(strict_types=1);
+
+
+namespace Rector\DowngradePhp80\Rector\ClassMethod;
+
+use PhpParser\Node;
+use PhpParser\Node\Stmt\ClassMethod;
+use Rector\Core\Rector\AbstractRector;
+use Rector\Core\ValueObject\Visibility;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @see \Rector\Tests\DowngradePhp80\Rector\ClassMethod\DowngradeAbstractPrivateMethodInTraitRector\DowngradeAbstractPrivateMethodInTraitRectorTest
+ */
+final class DowngradeAbstractPrivateMethodInTraitRector extends AbstractRector
+{
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Remove "abstract" from private methods in traits and adds an empty function body',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+trait SomeTrait
+{
+    abstract private function someAbstractPrivateFunction();
+}
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+trait SomeTrait
+{
+    private function someAbstractPrivateFunction() {}
+}
+CODE_SAMPLE
+                ),
+            ]
+        );
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [ClassMethod::class];
+    }
+
+    /**
+     * @param ClassMethod $node
+     */
+    public function refactor(Node $node): ?Node
+    {
```

Comment:
should it has "is trait" check?

Reply:
I think you can check something like this:

```php
        $scope = $classMethod->getAttribute(AttributeKey::SCOPE);
        if (! $scope instanceof Scope) {
            return null;
        }

        $classReflection = $scope->getClassReflection();
        if (! $classReflection instanceof ClassReflection) {
            return null;
        }

        if (! $classReflection->isTrait()) {
            return null;
        }
```        

or possibly use TraitAnalyzer : 

https://github.com/rectorphp/rector-src/blob/72df0f63c5015db2db86bf1b17a0d8393ad90911/rules/TypeDeclaration/NodeTypeAnalyzer/TraitTypeAnalyzer.php#L20

Reply:
It is impossible to have an `abstract private method` inside anything but a trait.
At least my PHPStorm is complaining like mad: [https://prnt.sc/1ki6eva](https://prnt.sc/1ki6eva)

And it even makes sense since private is limited to the same class, if we would put it inside of an abstract class there's no way to extend it, thus making an abstract private function in anything but a trait is impossible.

I can add a test case that covers abstract protected functions inside of an abstract class if you want though :)

Reply:
That's to avoid run this logic in class and interface, actually, you should can check `$node->getAttribute(AttributeKey::CLASS_NODE)` is Not an `Interface_` not a `Class_`

Reply:
or simpler: node->getAttribute(AttributeKey::PARENT) instanceof Trait_